src/lib/storage/buffer/frame.hpp

```diff
@@ -35,73 +46,129 @@ namespace hyrise {
 
-  // Removes a shared locked and returns true if the frame is now unlocked
+  // Decrement the shared latch. Returns true if the last shared latch has been released (state is UNLOCKED).
   bool unlock_shared();
 
+  // Unlocks the frame after an exclusive lock and sets the state to EVICTED.
+  void unlock_exclusive_and_set_evicted();
+
+  // Unlocks the frame and increments the version.
   void unlock_exclusive();
 
+  // Check if the frame is in state UNLOCKED.
   bool is_unlocked() const;
 
-  // State and version helper
+  // Load the complete state and version atomically.
   StateVersionType state_and_version() const;
-  static StateVersionType state(StateVersionType state_and_version);
-  static StateVersionType version(StateVersionType state_and_version);
-  static NodeID node_id(StateVersionType state_and_version);
 
-  void _debug_print();
+  // Extract the state from the atomic integer. The state is encoded in 16 bits.
+  static StateVersionType state(const StateVersionType state_and_version);
+
+  // Extract the version from the atomic integer. The version is encoded in 40 bits.
+  static StateVersionType version(const StateVersionType state_and_version);
+
+  // Extract the node id from the atomic integer. The node id is encoded in 7 bits.
+  static NodeID node_id(const StateVersionType state_and_version);
 
  private:
   // clang-format off
-  static constexpr uint64_t NODE_ID_MASK   = 0x00000F0000000000;
-  static constexpr uint64_t DIRTY_MASK       = 0x0000F00000000000;
-  static constexpr uint64_t STATE_MASK       = 0xFFFF000000000000;
-  static constexpr uint64_t VERSION_MASK     = 0x000000FFFFFFFFFF;
+
+  // The dirty flag is encoded in at bit 41.
+  static constexpr uint64_t _dirty_mask       = 0x0000800000000000;
+
+  // The NUMA is encoded 7 bits between the state an the the dirty mask.
+  static constexpr uint64_t _node_id_mask     = 0x00007F0000000000;
+
+  // The state is encoded in the upper 16 bits.
+  static constexpr uint64_t _state_mask       = 0xFFFF000000000000;
+
+  // The version is encoded in the lower 40 bits.
+  static constexpr uint64_t _version_mask     = 0x000000FFFFFFFFFF;
+
   // clang-format on
-  static_assert((NODE_ID_MASK ^ DIRTY_MASK ^ STATE_MASK ^ VERSION_MASK) == std::numeric_limits<StateVersionType>::max(),
-                "The given masks do not cover the whole StateVersionType");
 
-  static constexpr uint64_t NUM_BITS = sizeof(StateVersionType) * CHAR_BIT;
-  static constexpr uint64_t NODE_ID_SHIFT = std::countr_zero(NODE_ID_MASK);
-  static constexpr uint64_t DIRTY_SHIFT = std::countr_zero(DIRTY_MASK);
-  static constexpr uint64_t STATE_SHIFT = std::countr_zero(STATE_MASK);
+  static_assert((_node_id_mask ^ _dirty_mask ^ _state_mask ^ _version_mask) ==
+                    std::numeric_limits<StateVersionType>::max(),
+                "The given masks either overlap or do not cover the whole StateVersionType.");
+
+  // The number of bits for the state and version. It should be 64.
```

Comment:
> It should be 64.

Should we add a debug assertion for that somewhere?

Reply:
This is implicitly checked through the masks